### ISSUE TITLE
`assert_is_empty` must take size into consideration

### ISSUE THREAD

**Issue by nazar-pc** [rust-lang/rust-clippy#17523]
### Summary

I often use `vec.is_empty()` in tests when `vec` might contain a few megabytes of data. Replacing it with equality check is a very bad idea since it'll print many megabytes of text on failure.

### Lint Name

assert_is_empty

### Reproducer

I do not have a reduce example yet, but it is triggered, for example, on this line: https://github.com/nazar-pc/abundance/blob/184e6807a3138939f89a5ba18b550957421eacc1/crates/shared/ab-blake3/src/const_fn.rs#L153

```
warning: used `debug_assert!` to check that a value is empty
   --> crates/shared/ab-blake3/src/const_fn.rs:153:9
    |
153 |         debug_assert!(input.is_empty());
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#assert_is_empty
    = note: `-W clippy::assert-is-empty` implied by `-W clippy::pedantic`
    = help: to override `-W clippy::pedantic` add `#[allow(clippy::assert_is_empty)]`
help: use `debug_assert_eq!` to show the value on failure
    |
153 -         debug_assert!(input.is_empty());
153 +         debug_assert_eq!(input, []);
    |
```

But input can be gigabytes in size. I have even more tricky cases with reference counted containers and they seem to be even harder to recreate in reduced form. And they will be essentially impossible to check for actual size that is going to be printed in the end. So either this lint is mostly unusable or needs to be applied very conservatively.

### Version

```text
rustc 1.99.0-nightly (1a98b1e13 2026-08-07)
binary: rustc
commit-hash: 1a98b1e135b254f209c67d447b6d8bcd56a859e0
commit-date: 2026-08-07
host: x86_64-unknown-linux-gnu
release: 1.99.0-nightly
LLVM version: 23.1.0
```

**Comment by jaydaVis04** [comment]
Hi, I’m interested in working on this. I’d like to first reduce the reported case into a focused Clippy test, then inspect the conditions under which assert_is_empty suggests an equality assertion. My goal would be to make the lint conservative enough to avoid suggestions that could produce extremely large debug output while preserving useful diagnostics for simple cases. I’d also add UI tests for the behavior. Would this be a reasonable direction?

**Comment by Jarcho** [comment]
Same as your other issue not linting on byte slices should avoid enough cases.

---

That isn't the greatest example since removing `fill_buf` would remove any use for the `debug_assert` e.g.

```rust
const fn update(&mut self, mut input: &[u8]) -> &mut Self {
    if self.buf_len > 0 {
        let buf = self.buf.split_at_mut(self.buf_len as usize).1;
        let Some((head, tail)) = input.split_at_checked(buf.len()) else {
            buf.split_at_mut(input.len()).0.copy_from_slice(input);
            self.buf_len += input.len() as u8;
            debug_assert!(self.count() <= CHUNK_LEN);
            return self;
        };
        input = tail;
        buf.copy_from_slice(head);
        // compress self.buf
        self.buf_len = 0;
        self.buf = [0; _];
    }

    while let Some((head, tail)) = input.split_first_chunk::<BLOCK_LEN>() {
        input = tail;
        // compress block
    }

    self.buf.split_at_mut(input.len()).0.copy_from_slice(input);
    self.buf_len = input.len() as u8;
    debug_assert!(self.count() <= CHUNK_LEN);
    self
}
```
